contract_forecast/models/contract_line.py

```diff
@@ -69,18 +69,24 @@ def _generate_forecast_periods(self):
         for rec in self:
             rec.forecast_period_ids.unlink()
             if rec.recurring_next_date:
-                period_date_end = (
-                    rec.last_date_invoiced
-                    if rec.last_date_invoiced
-                    else rec.date_start - relativedelta(days=1)
-                )
+                period_date_start = rec.next_period_date_start
+                period_date_end = rec.next_period_date_end
+                recurring_next_date = rec.recurring_next_date
                 max_date_end = rec.date_end if not rec.is_auto_renew else False
                 while (
                     period_date_end
                     and rec._get_generate_forecast_periods_criteria(
                         period_date_end
                     )
                 ):
+                    if period_date_end and recurring_next_date:
+                        values.append(
+                            rec._prepare_contract_line_forecast_period(
```

Comment:
In general it's a bit awkward to have indented nested call like this compared to 
```
new_value = rec._prepare_contract_line_forecast_period(...)
values.append(new_value)
```
(and you win one line). But it's really a trivial detail.